src/cascadia/TerminalSettingsEditor/NullableColorPicker.cpp

```diff
@@ -0,0 +1,233 @@
+
+    void NullableColorPicker::_OnCurrentColorValueChanged(const DependencyObject& d, const DependencyPropertyChangedEventArgs& /*e*/)
+    {
+        const auto& obj{ d.try_as<Editor::NullableColorPicker>() };
+        get_self<NullableColorPicker>(obj)->_UpdateColorChips();
+    }
+
+    void NullableColorPicker::_UpdateColorChips()
+    {
+        const auto& currentColor = CurrentColor();
+        for (const auto& colorChip : _colorChips)
+        {
+            const auto& chipColor = colorChip.DataContext().as<Editor::ColorTableEntry>().Color();
+            colorChip.IsChecked(currentColor ?
+                                    chipColor.R == currentColor.Value().R && chipColor.G == currentColor.Value().G && chipColor.B == currentColor.Value().B :
+                                    false);
+        }
+    }
+
+    constexpr double _CalculateLuminance(const Windows::UI::Color& color)
+    {
+        auto linearize = [](float value) {
+            value /= 255.0;
+            return (value <= 0.03928) ? (value / 12.92) : (std::pow((value + 0.055) / 1.055, 2.4));
+        };
+        double r = linearize(color.R);
+        double g = linearize(color.G);
+        double b = linearize(color.B);
+        return 0.2126 * r + 0.7152 * g + 0.0722 * b;
+    }
+
+    SolidColorBrush NullableColorPicker::CalculateBorderBrush(const Windows::UI::Color& color)
+    {
+        // Accessibility requirement is a 3:1 contrast ratio
+        static const auto whiteLuminance = _CalculateLuminance(Colors::White());
+        const auto contrastRatio = (whiteLuminance + 0.05) / (_CalculateLuminance(color) + 0.05);
+        if (contrastRatio > 3.0)
+        {
+            return SolidColorBrush(Colors::White());
+        }
+        else
+        {
+            return SolidColorBrush(Colors::Black());
+        }
+    }
+
+    void NullableColorPicker::ColorChip_Clicked(const IInspectable& sender, const RoutedEventArgs& /*args*/)
+    {
+        const auto& btn = sender.as<Windows::UI::Xaml::Controls::Primitives::ToggleButton>();
+        const auto& colorEntryColor = btn.DataContext().as<Editor::ColorTableEntry>().Color();
+        const Microsoft::Terminal::Core::Color terminalColor{ colorEntryColor.R, colorEntryColor.G, colorEntryColor.B, colorEntryColor.A };
+        CurrentColor(terminalColor);
+        btn.IsChecked(true);
+    }
+
+    void NullableColorPicker::ColorChip_DataContextChanged(const IInspectable& sender, const DataContextChangedEventArgs& args)
+    {
+        if (const auto& toggleBtn = sender.try_as<Controls::Primitives::ToggleButton>())
+        {
+            if (const auto& currentColor = CurrentColor())
+            {
+                const auto& currentColorVal = currentColor.Value();
+                const auto& newChipColor = args.NewValue().as<Editor::ColorTableEntry>().Color();
+                toggleBtn.IsChecked(newChipColor.R == currentColorVal.R && newChipColor.G == currentColorVal.G && newChipColor.B == currentColorVal.B);
```

Comment:
Same here - even if you don't or do use `memcmp`, it may be worth writing a quick `static bool equalsColor(Windows::UI::Color a, Windows::UI::Color b)`.